spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java

```diff
@@ -0,0 +1,98 @@
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.S3ClientBuilder;
+import software.amazon.awssdk.services.s3.S3Configuration;
+
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.util.StringUtils;
+
+/**
+ * {@link EnableAutoConfiguration} for {@link S3Client} and {@link S3ProtocolResolver}.
+ *
+ * @author Maciej Walkowiak
+ */
+@ConditionalOnClass({ S3Client.class, CrossRegionS3Client.class })
+@EnableConfigurationProperties(S3Properties.class)
+@Configuration(proxyBeanMethods = false)
+@Import(S3ProtocolResolver.class)
+public class S3AutoConfiguration {
+
+	private final S3Properties properties;
+
+	public S3AutoConfiguration(S3Properties properties) {
+		this.properties = properties;
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	S3ClientBuilder s3ClientBuilder(AwsCredentialsProvider credentialsProvider, AwsRegionProvider regionProvider) {
+		Region region = StringUtils.hasLength(this.properties.getRegion()) ? Region.of(this.properties.getRegion())
+				: regionProvider.getRegion();
+		S3ClientBuilder builder = S3Client.builder().credentialsProvider(credentialsProvider).region(region)
+				.overrideConfiguration(SpringCloudClientConfiguration.clientOverrideConfiguration())
+				.serviceConfiguration(s3ServiceConfiguration());
+		Optional.ofNullable(this.properties.getEndpoint()).ifPresent(builder::endpointOverride);
+		return builder;
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	S3Client s3Client(S3ClientBuilder s3ClientBuilder) {
+		return new CrossRegionS3Client(s3ClientBuilder);
+	}
+
+	private S3Configuration s3ServiceConfiguration() {
+		S3Configuration.Builder config = S3Configuration.builder();
+		if (properties.getAccelerateModeEnabled() != null) {
+			config.accelerateModeEnabled(properties.getAccelerateModeEnabled());
+		}
+		if (properties.getChecksumValidationEnabled() != null) {
+			config.checksumValidationEnabled(properties.getChecksumValidationEnabled());
+		}
+		if (properties.getChunkedEncodingEnabled() != null) {
+			config.chunkedEncodingEnabled(properties.getChunkedEncodingEnabled());
+		}
+		if (properties.getDualstackEnabled() != null) {
+			config.dualstackEnabled(properties.getDualstackEnabled());
+		}
+		if (properties.getPathStyleAccessEnabled() != null) {
+			config.pathStyleAccessEnabled(properties.getPathStyleAccessEnabled());
+		}
+		if (properties.getUseArnRegionEnabled() != null) {
+			config.useArnRegionEnabled(properties.getUseArnRegionEnabled());
+		}
```

Comment:
We can use [PropertyMapper](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/context/properties/PropertyMapper.html) here